MapboxCoreNavigation/RouteController.swift

```diff
@@ -472,3 +618,134 @@ extension RouteController: CLLocationManagerDelegate {
+            outstandingFeedbackEvents.remove(at: index)
+        }
+        
+        let eventName = event.eventDictionary["event"] as! String
+        
+        event.eventDictionary["locationsBefore"] = sessionState.pastLocations.allObjects.filter {$0.timestamp <= event.timestamp}.map {$0.eventDictionary}
+        event.eventDictionary["locationsAfter"] = sessionState.pastLocations.allObjects.filter {$0.timestamp > event.timestamp}.map {$0.eventDictionary}
+        
+        events.enqueueEvent(withName: eventName, attributes: event.eventDictionary)
+        events.flush()
+    }
+
+    func sendArriveEvent() {
+        let eventName = "navigation.arrive"
+        
+        var eventDictionary = events.addDefaultEvents(routeController: self)
+        eventDictionary["event"] = eventName
+        
+        events.enqueueEvent(withName: eventName, attributes: eventDictionary)
+        events.flush()
+    }
+    
+    func sendCancelEvent() {
+        let eventName = "navigation.cancel"
+
+        var eventDictionary = events.addDefaultEvents(routeController: self)
+        eventDictionary["event"] = eventName
+        eventDictionary["arrivalTimestamp"] = sessionState.arrivalTimestamp?.ISO8601 ?? NSNull()
+
+        events.enqueueEvent(withName: eventName, attributes: eventDictionary)
+        events.flush()
+    }
+    
+    func enqueueFeedbackEvent(type: FeedbackType, description: String?) {
+        let eventName = "navigation.feedback"
+        
+        var eventDictionary = events.addDefaultEvents(routeController: self)
+        eventDictionary["event"] = eventName
+        
+        eventDictionary["userId"] = UIDevice.current.identifierForVendor?.uuidString
+        eventDictionary["feedbackType"] = type.rawValue
+        eventDictionary["description"] = description
+        for (key, value) in routeProgress.currentLegProgress.upcomingManeuverDictionary {
+            eventDictionary["step\(key.uppercaseFirst)"] = value
+        }
+        // FIXME: Replace the above enumeration with the below once https://github.com/mapbox/api-events/issues/265 is fixed
+        // eventDictionary["upcomingStep"] = routeProgress.currentLegProgress.upcomingManeuverDictionary
+        eventDictionary["screenshot"] = captureScreen(scaledToFit: 250)?.base64EncodedString()
+
+        outstandingFeedbackEvents.append(FeedbackEvent(timestamp: Date(), eventDictionary: eventDictionary))
+    }
+    
+    func enqueueRerouteEvent() {
+        let eventName = "navigation.reroute"
+
+        let timestamp = Date()
+        
+        var eventDictionary = events.addDefaultEvents(routeController: self)
+        eventDictionary["event"] = eventName
+        
+        eventDictionary["secondsSinceLastReroute"] = sessionState.lastReroute != nil ? round(timestamp.timeIntervalSince(sessionState.lastReroute!)) : -1
+        for (key, value) in routeProgress.currentLegProgress.upcomingManeuverDictionary {
+            eventDictionary["step\(key.uppercaseFirst)"] = value
```

Comment:
As long as there aren’t any spaces in the keys, `String.capitalized` does the same thing.